providers/implementations/serializers/rsa_print.c

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/core_numbers.h>
+#include <openssl/core_names.h>
+#include <openssl/bn.h>
+#include <openssl/rsa.h>
+#include <openssl/types.h>
+#include <openssl/safestack.h>
+#include "prov/bio.h"
+#include "prov/implementations.h"
+#include "crypto/rsa.h"
+#include "crypto/bn.h"
+
+DEFINE_SPECIAL_STACK_OF_CONST(BIGNUM_const, BIGNUM)
+
+static OSSL_OP_serializer_serialize_data_fn print_priv_rsa_data;
+static OSSL_OP_serializer_serialize_object_fn print_priv_rsa;
+static OSSL_OP_serializer_serialize_data_fn print_pub_rsa_data;
+static OSSL_OP_serializer_serialize_object_fn print_pub_rsa;
+
+static OSSL_OP_keymgmt_importkey_fn *get_rsa_importkey(void)
+{
+    static OSSL_OP_keymgmt_importkey_fn *rsa_importkey = NULL;
+    const OSSL_DISPATCH *fns = rsa_keymgmt_functions;
+
+    /* Pilfer the RSA keymgmt */
+    if (rsa_importkey == NULL) {
+        for (; fns->function_id != 0; fns++)
+            if (fns->function_id == OSSL_FUNC_KEYMGMT_IMPORTKEY) {
+                rsa_importkey = OSSL_get_OP_keymgmt_importkey(fns);
+                break;
+            }
+    }
+
+    return rsa_importkey;
+}
+
+# ifdef SIXTY_FOUR_BIT_LONG
+#  define BN_FMTu "%lu"
+#  define BN_FMTx "%lx"
+# endif
+
+# ifdef SIXTY_FOUR_BIT
+#  define BN_FMTu "%llu"
+#  define BN_FMTx "%llx"
+# endif
+
+# ifdef THIRTY_TWO_BIT
+#  define BN_FMTu "%u"
+#  define BN_FMTx "%x"
+# endif
+
+static int print_labeled_bignum(BIO *out, const char *label, const BIGNUM *n)
+{
+    const char *neg;
+    const char *post_label_spc = " ";
+    int bytes;
+    BN_ULONG *words;
+    int i, off;
+
+    if (n == NULL)
+        return 0;
+    if (label == NULL) {
+        label = "";
+        post_label_spc = "";
+    }
+
+    bytes = BN_num_bytes(n);
+    words = bn_get_words(n);
+    neg = BN_is_negative(n) ? "-" : "";
+
+    if (BN_is_zero(n))
+        return ossl_prov_bio_printf(out, "%s%s0\n", label, post_label_spc);
+
+#if 0
+    if (BN_num_bytes(n) <= BN_BYTES)
+        return ossl_prov_bio_printf(out,
+                                    "%s%s%s" BN_FMTu " (%s0x" BN_FMTx ")\n",
+                                    label, post_label_spc, neg, words[0],
+                                    neg, words[0]);
+#endif
+
+    if (neg[0] == '-')
+        neg = " (Negative)";
```

Comment:
RSA values are never negative. Or do you plan for this to become a more generic function? In which case it doesn't belong in this file.

Reply:
Granted, I basically copied what is done downcall from `ASN1_bn_print`.  Are you *absolutely* sure we will never have a negative large number?  Not even in parameters such as PSS, or any other future set of params we don't know about?

Reply:
From a quick read `crypto/rsa/rsa_ameth.c`, it seems that the PSS saltlen can have negative values...  seeing that this comes in form of a `ASN1_INTEGER`, which translates into an BIGNUM, I think it may be wise to keep the negative number support...